packages/authorization-process/src/services/authorizationService.ts

```diff
@@ -384,6 +440,85 @@ export function authorizationServiceBuilder(
         showUsers: updatedClient.consumerId === authData.organizationId,
       };
     },
+    async addClientPurpose({
+      clientId,
+      seed,
+      organizationId,
+      correlationId,
+      logger,
+    }: {
+      clientId: ClientId;
+      seed: ApiPurposeAdditionSeed;
+      organizationId: TenantId;
+      correlationId: string;
+      logger: Logger;
+    }): Promise<void> {
+      logger.info(
+        `Adding purpose with id ${seed.purposeId} to client ${clientId}`
+      );
+      const purposeId: PurposeId = unsafeBrandId(seed.purposeId);
+
+      const client = await retrieveClient(clientId, readModelService);
+      assertOrganizationIsClientConsumer(organizationId, client.data);
+
+      const purpose = await retrievePurpose(purposeId, readModelService);
+      assertOrganizationIsPurposeConsumer(organizationId, purpose);
+
+      const eservice = await retrieveEService(
+        purpose.eserviceId,
+        readModelService
+      );
+
+      const agreements = await readModelService.getAgreements(
+        eservice.id,
+        organizationId
+      );
+      const agreement = agreements
+        .filter(
+          (a) =>
+            a.state === agreementState.active ||
+            a.state === agreementState.suspended
+        )
+        .sort((a1, a2) => a1.createdAt.getTime() - a2.createdAt.getTime())[0];
+
+      if (agreement === undefined) {
+        throw agreementNotFound(eservice.id, organizationId);
+      }
+
+      retrieveDescriptor(agreement.descriptorId, eservice);
+
+      const invalidPurposeVersionStates: PurposeVersionState[] = [
+        purposeVersionState.archived,
+        purposeVersionState.rejected,
+        purposeVersionState.draft,
+        purposeVersionState.waitingForApproval,
+      ];
+      const purposeVersion = purpose.versions.find(
+        (v) => !invalidPurposeVersionStates.includes(v.state)
+      );
```

Comment:
```suggestion
      const invalidPurposeVersionStates: Set<PurposeVersionState> = new Set([
        purposeVersionState.archived,
        purposeVersionState.rejected,
        purposeVersionState.draft,
        purposeVersionState.waitingForApproval,
      ]);
      const purposeVersion = purpose.versions.find(
        (v) => !invalidPurposeVersionStates.has(v.state)
      );
```
I think that a `Set` in this case would perform better than an `Array`

Reply:
Updated